controller/src/main/java/org/jboss/as/controller/CapabilityReferenceRecorder.java

```diff
@@ -186,34 +232,42 @@ public final void addCapabilityRequirements(OperationContext context, String att
          * @throws AssertionError if the requirements discussed in the class javadoc are not fulfilled
          */
         @Override
-        public final void removeCapabilityRequirements(OperationContext context, String attributeName, String... attributeValues) {
-            processCapabilityRequirement(context, attributeName, true, attributeValues);
+        public final void removeCapabilityRequirements(OperationContext context, Resource resource, String attributeName, String... attributeValues) {
+            processCapabilityRequirement(context, resource, attributeName, true, attributeValues);
         }
 
-        private void processCapabilityRequirement(OperationContext context, String attributeName, boolean remove, String... attributeValues) {
-            String dependentName = getDependentName(context);
-
+        protected void processCapabilityRequirement(OperationContext context, Resource resource, String attributeName, boolean remove, String... attributeValues) {
+            RuntimeCapability cap = getDependentCapability(context);
+            String dependentName = getDependentName(cap, context);
             for (String attributeValue : attributeValues) {
-                String requirementName = RuntimeCapability.buildDynamicCapabilityName(baseRequirementName, attributeValue);
-                if (remove) {
-                    context.deregisterCapabilityRequirement(requirementName, dependentName);
-                } else {
-                    context.registerAdditionalCapabilityRequirement(requirementName, dependentName, attributeName);
+                if (attributeValue != null || !cap.getDynamicOptionalRequirements().contains(baseRequirementName)) {
```

Comment:
Currently there are few uses of registering optional dynamic requirements.
It was just that this info was nowhere to be used.

By definition all requirements are mandatory, so if you have an optional requirement which is in some cases a design requirement, you need to express this in some way.

With this change, what happens is that if attribute is null aka is not configured and requirement is not specified as "optional" requirement as such must be added to the registry.

The requirement name becomes bit bogus, as it is basename + "null" but none the less it tells the subsystem author something is wrong.

Going forward whole validation for such scenarios should be improved, but that belongs into validation of capabilities not gathering the requirements.